extensions/omni_vfs/migrate/4_table_fs.sql

```diff
@@ -0,0 +1,217 @@
+    update of id, filesystem_id, filepath, kind
+    on table_fs_files
+    for each row
+execute function table_fs_files_trigger();
+
+
+-- TODO: refactor this into [sparse] [reusable] chunks
+create table table_fs_file_data
+(
+    file bigint primary key references table_fs_files (id),
+    data bytea not null
+);
+
+/*  
+    external storage optimizes fetching parts of data
+    at the expense of increased storage due to disabled compression
+*/
+alter table omni_vfs.table_fs_file_data alter column data set storage external;
+
+create or replace function table_fs_file_data_trigger() returns trigger as
+$$
+begin
+    update omni_vfs.table_fs_files
+    set
+        created_at = (case tg_op when 'INSERT' then statement_timestamp() else created_at end),
+        accessed_at = statement_timestamp(),
+        modified_at = statement_timestamp()
+    where
+        id = new.file;
+    return new;
+end;
+$$ language plpgsql;
+
+create trigger table_fs_file_data_trigger
+    before insert or update
+    on table_fs_file_data
+    for each row
+execute function table_fs_file_data_trigger();
+
+create type table_fs as
+(
+    id integer
+);
+
+create function table_fs(filesystem_name text) returns omni_vfs.table_fs
+    language plpgsql
+as
+$$
+declare
+    result_id integer;
+begin
+    select
+        id
+    from
+        omni_vfs.table_fs_filesystems
+    where
+        table_fs_filesystems.name = filesystem_name
+    into result_id;
+    if not found then
+        insert
+        into
+            omni_vfs.table_fs_filesystems (name)
+        values (filesystem_name)
+        returning id into result_id;
+    end if;
+    return row (result_id);
+end
+$$;
+
+create function list(fs table_fs, path text, fail_unpermitted boolean default true) returns setof omni_vfs_types_v1.file
+    stable
+    language sql
+as
+$$
+with
+    search(path) as (
+        select
+            case
+                when omni_vfs.canonicalize_path(path, absolute => true) = '/' then ''
+                else omni_vfs.canonicalize_path(path, absolute => true)
+            end
+    )
+select
+    distinct case
+        -- filepath exactly matches search.path
+        when filepath = search.path then row(omni_vfs.basename(filepath), kind)::omni_vfs_types_v1.file
+        -- search.path is prefix of filepath
+        else (
+                case
+                    -- only one '/' left, eg. /file
+                    when split_part(substr(filepath, length(search.path)+1), '/', 3) = '' then
+                        row(split_part(substr(filepath, length(search.path)+1), '/', 2), kind)::omni_vfs_types_v1.file
+                    -- more than one '/', eg. /dir/file
+                    else row(split_part(substr(filepath, length(search.path)+1), '/', 2), 'dir')::omni_vfs_types_v1.file
+                end
+        )
+    end
+from omni_vfs.table_fs_files, search
+where filesystem_id = fs.id and
+(filepath like (search.path || '/%') or
```

Comment:
using `pg_trgm` to create index on `filepath` should help with `like` operator but listing of paths such `/` will most likely lead to seq scan because we store entire filepath in each entry, leading to every entry matching(postgres planner may avoid using indexes in such cases)  just to produce few directories and files that are present in root directory

Reply:
It doesn't really have to be a `like` operator but just substring, really. Does this help us?

Aside, I wonder if (at least theoretically, as this is beyond the scope of this PR) we can implement a custom index.

Reply:
Yes, it can be converted to use substring but the issue still remains that postgres planner sees too many entries are matching for certain entries so it will opt for seq scan.

It would have been ok if most of the entries in the root directory are different but filesystems usually have some small number of directories in root. 

For eg. if the following files are in the table `/dir/file1`, `/dir/file2`, `/dir/file3`, we end up processing all the three entries just to produce `dir` when listing `/`